File: scripts/update-readme.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";

const README_PATH = path.resolve(process.env.PROFILE_README_PATH ?? "README.md");
const PAGES_OWNER = process.env.PAGES_OWNER ?? "ankhseraph";
const PAGES_REPO = process.env.PAGES_REPO ?? "pages";
const PAGES_BRANCH = process.env.PAGES_BRANCH ?? "main";
const CODEBERG_API_BASE = (process.env.CODEBERG_API_BASE ?? "https://codeberg.org/api/v1").replace(/\/+$/, "");
const PAGES_BASE_URL = (process.env.PAGES_BASE_URL ?? "https://ankhseraph.codeberg.page").replace(/\/+$/, "");

const RAW_BASE = `https://codeberg.org/${PAGES_OWNER}/${PAGES_REPO}/raw/branch/${PAGES_BRANCH}`;
const LATEST_PROJECTS_COUNT = Number(process.env.LATEST_PROJECTS_COUNT ?? "2");
const LATEST_LOGBOOK_COUNT = Number(process.env.LATEST_LOGBOOK_COUNT ?? "2");

function isIsoDateMd(name) {
  return /^\d{4}-\d{2}-\d{2}\.md$/.test(String(name));
}

function isHiddenLike(name) {
  const lower = String(name).toLowerCase();
  return (
    name === "EXAMPLE.md" ||
    lower.startsWith("_") ||
    lower.startsWith("example") ||
    lower.startsWith("sample") ||
    lower.startsWith("template")
  );
}

function extractFirstParagraph(markdown) {
  const lines = String(markdown).split(/\r?\n/);
  let index = 0;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  if (index < lines.length && lines[index].startsWith("#")) index += 1;
  while (index < lines.length && lines[index].trim() === "") index += 1;

  const paragraph = [];
  while (index < lines.length) {
    const line = lines[index];
    if (line.trim() === "") break;
    if (line.startsWith("#") || line.startsWith("```") || line.startsWith("- ") || line.startsWith(">")) break;
    paragraph.push(line.trim());
    index += 1;
  }
  return paragraph.join(" ").trim();
}

function stripMarkdown(text) {
  return String(text)
    .replace(/!\[([^\]]*)\]\([^)]+\)/g, "$1")
    .replace(/\[([^\]]+)\]\([^)]+\)/g, "$1")
    .replace(/`([^`]+)`/g, "$1")
    .replace(/\*\*([^*]+)\*\*/g, "$1")
    .replace(/\*([^*]+)\*/g, "$1")
    .replace(/_([^_]+)_/g, "$1")
    .replace(/https?:\/\/\S+/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

function toOneSentence(text, maxLen = 220) {
  const cleaned = stripMarkdown(text);
  if (!cleaned) return "";
  const end = cleaned.search(/[.!?](\s|$)/);
  let sentence = end === -1 ? cleaned : cleaned.slice(0, end + 1);
  if (sentence.length > maxLen) sentence = `${sentence.slice(0, maxLen - 1).trimEnd()}…`;
  return sentence;
}

function extractH1(markdown) {
  const lines = String(markdown).split(/\r?\n/);
  for (const line of lines) {
    if (line.trim().startsWith("# ")) return line.trim().slice(2).trim();
  }
  return "";
}

function extractLogbookSubtitle(markdown) {
  const lines = String(markdown).split(/\r?\n/);
  let index = 0;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  if (index < lines.length && lines[index].trim().startsWith("#")) index += 1;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  while (index < lines.length) {
    const line = lines[index].trim();
    if (!line) break;
    if (line.startsWith("#") || line.startsWith("```") || line.startsWith("- ") || line.startsWith(">")) break;
    return line;
  }
  return "";
}

function extractLogbookBodyExcerpt(markdown) {
  const lines = String(markdown).split(/\r?\n/);
  let index = 0;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  if (index < lines.length && lines[index].trim().startsWith("#")) index += 1;
  while (index < lines.length && lines[index].trim() === "") index += 1;
  if (index < lines.length && !lines[index].trim().startsWith("#")) index += 1; // subtitle line
  while (index < lines.length) {
    const line = lines[index].trim();
    if (!line) {
      index += 1;
      continue;
    }
    if (line.startsWith("#") || line.startsWith("```") || line.startsWith("- ") || line.startsWith(">")) {
      index += 1;
      continue;
    }
    break;
  }

  const paragraph = [];
  while (index < lines.length) {
    const line = lines[index];
    if (line.trim() === "") break;
    if (line.startsWith("#") || line.startsWith("```") || line.startsWith("- ") || line.startsWith(">")) break;
    paragraph.push(line.trim());
    index += 1;
  }
  return paragraph.join(" ").trim();
}

function htmlEscape(text) {
  return String(text)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/\"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function linkToPagesMd(relPath) {
  const encoded = encodeURIComponent(relPath.replace(/\\/g, "/"));
  return `${PAGES_BASE_URL}/md.html?file=${encoded}`;
}

function pluralize(n, singular, plural = `${singular}s`) {
  return n === 1 ? singular : plural;
}

function renderStrongLink({ title, href }) {
  const safeTitle = htmlEscape(title);
  if (!href) return `<strong>${safeTitle}</strong>`;
  return `<strong><a href="${htmlEscape(href)}">${safeTitle}</a></strong>`;
}

function renderLinksList(items) {
  if (!items || items.length === 0) return `<div class="muted">None yet.</div>`;
  const lis = items
    .map((item) => `<li><a href="${htmlEscape(item.href)}">${htmlEscape(item.title)}</a></li>`)
    .join("");
  return `<ul>${lis}</ul>`;
}

function updateReadmeGeneratedBlock(newBlock) {
  const readme = fs.readFileSync(README_PATH, "utf8");
  const start = "<!-- AUTO-GENERATED:START";
  const end = "<!-- AUTO-GENERATED:END -->";
  const startIndex = readme.indexOf(start);
  const endIndex = readme.indexOf(end);

  if (startIndex === -1 || endIndex === -1 || endIndex < startIndex) {
    throw new Error(`Could not find AUTO-GENERATED markers in ${README_PATH}`);
  }

  const before = readme.slice(0, startIndex);
  const after = readme.slice(endIndex + end.length);
  fs.writeFileSync(README_PATH, `${before}${newBlock}${after}`, "utf8");
}

async function fetchJson(url) {
  const response = await fetch(url, { headers: { "cache-control": "no-store" } });
  if (!response.ok) throw new Error(`HTTP ${response.status} for ${url}`);
  return response.json();
}

async function fetchText(url) {
  const response = await fetch(url, { headers: { "cache-control": "no-store" } });
  if (!response.ok) throw new Error(`HTTP ${response.status} for ${url}`);
  return response.text();
}

async function latestCommitDateForPath(repoPath) {
  const query = new URLSearchParams({ path: repoPath, limit: "1" });
  const url = `${CODEBERG_API_BASE}/repos/${PAGES_OWNER}/${PAGES_REPO}/commits?${query.toString()}`;
  const commits = await fetchJson(url);
  const dateString = commits?.[0]?.commit?.committer?.date;
  const date = dateString ? new Date(dateString) : null;
  return Number.isFinite(date?.valueOf()) ? date : null;
}

async function main() {
  const projectsIndex = await fetchJson(`${RAW_BASE}/projects/index.json`).catch(() => []);
  const logbookIndex = await fetchJson(`${RAW_BASE}/logbook/index.json`).catch(() => []);

  const logbookFiles = Array.isArray(logbookIndex) ? logbookIndex.filter((name) => isIsoDateMd(name)) : [];
  logbookFiles.sort((a, b) => String(b).localeCompare(String(a)));

  const projectFiles = Array.isArray(projectsIndex)
    ? projectsIndex
        .filter((name) => String(name).toLowerCase().endsWith(".md"))
        .filter((name) => !isHiddenLike(name))
    : [];

  const projectsCount = projectFiles.length;
  const logbookCount = logbookFiles.length;
  const totalWriteups = projectsCount + logbookCount;

  const latestLogbookFile = logbookFiles[0] ?? null;
  const projectCandidates = await (async () => {
    if (projectFiles.length === 0) return [];

    const candidates = await Promise.all(
      projectFiles.map(async (name) => {
        const repoPath = `projects/${name}`;
        const date = await latestCommitDateForPath(repoPath).catch(() => null);
        return { name, date };
      }),
    );

    candidates.sort((a, b) => {
      const ad = a.date?.valueOf() ?? 0;
      const bd = b.date?.valueOf() ?? 0;
      return bd - ad || String(b.name).localeCompare(String(a.name));
    });

    return candidates;
  })();

  const latestProjectFile = projectCandidates[0]?.name ?? null;

  let latestProjectTitle = "No projects yet";
  let latestProjectDesc = "";
  let latestProjectLink = "";

  if (latestProjectFile) {
    const rel = `projects/${latestProjectFile}`;
    const md = await fetchText(`${RAW_BASE}/${rel}`);
    latestProjectTitle = extractH1(md) || latestProjectFile.replace(/\.md$/i, "");
    latestProjectDesc = toOneSentence(extractFirstParagraph(md));
    latestProjectLink = linkToPagesMd(rel);
  }

  let latestLogbookTitle = "No logbook entries yet";
  let latestLogbookDesc = "";
  let latestLogbookLink = "";

  if (latestLogbookFile) {
    const rel = `logbook/${latestLogbookFile}`;
    const md = await fetchText(`${RAW_BASE}/${rel}`);
    const date = latestLogbookFile.replace(/\.md$/i, "");
    const subtitle = extractLogbookSubtitle(md);
    latestLogbookTitle = subtitle ? `${date} — ${subtitle}` : date;
    latestLogbookDesc = toOneSentence(extractLogbookBodyExcerpt(md));
    latestLogbookLink = linkToPagesMd(rel);
  }

  const latestProjects = await Promise.all(
    projectCandidates.slice(0, Math.max(0, LATEST_PROJECTS_COUNT)).map(async ({ name }) => {
      const rel = `projects/${name}`;
      const md = await fetchText(`${RAW_BASE}/${rel}`).catch(() => "");
      const title = extractH1(md) || name.replace(/\.md$/i, "");
      return { title, href: linkToPagesMd(rel) };
    }),
  );

  const latestLogbook = await Promise.all(
    logbookFiles.slice(0, Math.max(0, LATEST_LOGBOOK_COUNT)).map(async (name) => {
      const rel = `logbook/${name}`;
      const md = await fetchText(`${RAW_BASE}/${rel}`).catch(() => "");
      const date = name.replace(/\.md$/i, "");
      const subtitle = extractLogbookSubtitle(md);
      const title = subtitle ? `${date} — ${subtitle}` : date;
      return { title, href: linkToPagesMd(rel) };
    }),
  );

  const newBlock = `<!-- AUTO-GENERATED:START (do not edit by hand) -->
<div align="center">
  <h2>Total writeups</h2>
  <h1>${totalWriteups}</h1>
  <p>${projectsCount} ${pluralize(projectsCount, "project")} • ${logbookCount} logbook ${pluralize(logbookCount, "entry", "entries")}</p>
</div>

---

## Latest highlights

<table>
  <tr>
    <td width="50%">
      <h3>Latest project writeup</h3>
      <p>${renderStrongLink({ title: latestProjectTitle, href: latestProjectLink })}</p>
      <p>${htmlEscape(latestProjectDesc)}</p>
    </td>
    <td width="50%">
      <h3>Latest logbook entry</h3>
      <p>${renderStrongLink({ title: latestLogbookTitle, href: latestLogbookLink })}</p>
      <p>${htmlEscape(latestLogbookDesc)}</p>
    </td>
  </tr>
</table>

<table>
  <tr>
    <td width="50%">
      <h3>Latest projects</h3>
      ${renderLinksList(latestProjects)}
    </td>
    <td width="50%">
      <h3>Latest blog (logbook)</h3>
      ${renderLinksList(latestLogbook)}
    </td>
  </tr>
</table>

<p align="center"><small>Auto-updated from my <a href="https://codeberg.org/ankhseraph/pages">pages repo</a>.</small></p>
<!-- AUTO-GENERATED:END -->`;

  updateReadmeGeneratedBlock(newBlock);
}

try {
  await main();
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
}

File: scripts/update-readme.py
#!/usr/bin/env python3
import html
import json
import os
import re
import subprocess
import urllib.parse
import urllib.request
from datetime import datetime, timezone
from pathlib import Path

README_PATH = Path(os.environ.get("PROFILE_README_PATH", "README.md")).resolve()
PAGES_OWNER = os.environ.get("PAGES_OWNER", "ankhseraph")
PAGES_REPO = os.environ.get("PAGES_REPO", "pages")
PAGES_BRANCH = os.environ.get("PAGES_BRANCH", "main")
CODEBERG_API_BASE = os.environ.get("CODEBERG_API_BASE", "https://codeberg.org/api/v1").rstrip("/")
PAGES_BASE_URL = os.environ.get("PAGES_BASE_URL", "https://ankhseraph.codeberg.page").rstrip("/")
PROJECTS_DIR = os.environ.get("PROJECTS_DIR", "projects")
LOGBOOK_DIR = os.environ.get("LOGBOOK_DIR", "logbook")
_env_local = os.environ.get("PAGES_LOCAL_PATH", "")
PAGES_LOCAL_PATH = _env_local or (str(Path.home() / "pages") if (Path.home() / "pages").exists() else "")

RAW_BASE = f"https://codeberg.org/{PAGES_OWNER}/{PAGES_REPO}/raw/branch/{PAGES_BRANCH}"


def fetch_json(url: str):
    req = urllib.request.Request(url, headers={"cache-control": "no-store"})
    with urllib.request.urlopen(req, timeout=20) as resp:
        return json.load(resp)


def fetch_text(url: str) -> str:
    req = urllib.request.Request(url, headers={"cache-control": "no-store"})
    with urllib.request.urlopen(req, timeout=20) as resp:
        return resp.read().decode("utf-8", errors="replace")

def read_local_text(base: Path, rel_path: str) -> str:
    return (base / rel_path).read_text(encoding="utf-8")

def read_local_json(base: Path, rel_path: str):
    return json.loads((base / rel_path).read_text(encoding="utf-8"))


def is_iso_date_md(name: str) -> bool:
    return bool(re.fullmatch(r"\d{4}-\d{2}-\d{2}\.md", str(name)))


def is_hidden_like(name: str) -> bool:
    lower = str(name).lower()
    return (
        name == "EXAMPLE.md"
        or lower.startswith("_")
        or lower.startswith("example")
        or lower.startswith("sample")
        or lower.startswith("template")
    )


def extract_first_paragraph(markdown: str) -> str:
    lines = str(markdown).splitlines()
    idx = 0
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    if idx < len(lines) and lines[idx].startswith("#"):
        idx += 1
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    paragraph = []
    while idx < len(lines):
        line = lines[idx]
        if line.strip() == "":
            break
        if line.startswith("#") or line.startswith("```") or line.startswith("- ") or line.startswith(">"):
            break
        paragraph.append(line.strip())
        idx += 1
    return " ".join(paragraph).strip()


def strip_markdown(text: str) -> str:
    text = re.sub(r"!\[([^\]]*)\]\([^)]+\)", r"\1", text)
    text = re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", text)
    text = re.sub(r"`([^`]+)`", r"\1", text)
    text = re.sub(r"\*\*([^*]+)\*\*", r"\1", text)
    text = re.sub(r"\*([^*]+)\*", r"\1", text)
    text = re.sub(r"_([^_]+)_", r"\1", text)
    text = re.sub(r"https?://\S+", "", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def to_one_sentence(text: str, max_len: int = 220) -> str:
    cleaned = strip_markdown(text)
    if not cleaned:
        return ""
    match = re.search(r"[.!?](\s|$)", cleaned)
    sentence = cleaned if not match else cleaned[: match.end()]
    if len(sentence) > max_len:
        sentence = sentence[: max_len - 1].rstrip() + "…"
    return sentence


def extract_h1(markdown: str) -> str:
    for line in str(markdown).splitlines():
        if line.strip().startswith("# "):
            return line.strip()[2:].strip()
    return ""


def extract_logbook_subtitle(markdown: str) -> str:
    lines = str(markdown).splitlines()
    idx = 0
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    if idx < len(lines) and lines[idx].strip().startswith("#"):
        idx += 1
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    while idx < len(lines):
        line = lines[idx].strip()
        if not line:
            break
        if line.startswith("#") or line.startswith("```") or line.startswith("- ") or line.startswith(">"):
            break
        return line
    return ""


def extract_logbook_body_excerpt(markdown: str) -> str:
    lines = str(markdown).splitlines()
    idx = 0
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    if idx < len(lines) and lines[idx].strip().startswith("#"):
        idx += 1
    while idx < len(lines) and lines[idx].strip() == "":
        idx += 1
    if idx < len(lines) and not lines[idx].strip().startswith("#"):
        idx += 1
    while idx < len(lines):
        line = lines[idx].strip()
        if not line:
            idx += 1
            continue
        if line.startswith("#") or line.startswith("```") or line.startswith("- ") or line.startswith(">"):
            idx += 1
            continue
        break
    paragraph = []
    while idx < len(lines):
        line = lines[idx]
        if line.strip() == "":
            break
        if line.startswith("#") or line.startswith("```") or line.startswith("- ") or line.startswith(">"):
            break
        paragraph.append(line.strip())
        idx += 1
    return " ".join(paragraph).strip()


def link_to_pages_md(rel_path: str) -> str:
    encoded = urllib.parse.quote(rel_path.replace("\\", "/"))
    return f"{PAGES_BASE_URL}/md.html?file={encoded}"


def render_strong_link(title: str, href: str) -> str:
    safe_title = html.escape(title)
    if not href:
        return f"<strong>{safe_title}</strong>"
    return f"<strong><a href=\"{html.escape(href)}\">{safe_title}</a></strong>"


def update_readme_generated_block(new_block: str) -> None:
    readme = README_PATH.read_text(encoding="utf-8")
    start = "<!-- AUTO-GENERATED:START"
    end = "<!-- AUTO-GENERATED:END -->"
    start_idx = readme.find(start)
    end_idx = readme.find(end)
    if start_idx == -1 or end_idx == -1 or end_idx < start_idx:
        raise RuntimeError(f"Could not find AUTO-GENERATED markers in {README_PATH}")
    before = readme[:start_idx]
    after = readme[end_idx + len(end) :]
    README_PATH.write_text(f"{before}{new_block}{after}", encoding="utf-8")


def latest_commit_date_for_path(repo_path: str):
    query = urllib.parse.urlencode({"path": repo_path, "limit": "1"})
    url = f"{CODEBERG_API_BASE}/repos/{PAGES_OWNER}/{PAGES_REPO}/commits?{query}"
    commits = fetch_json(url)
    date_string = None
    if isinstance(commits, list) and commits:
        date_string = commits[0].get("commit", {}).get("committer", {}).get("date")
    if not date_string:
        return None
    try:
        dt = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
        return dt.astimezone(timezone.utc)
    except ValueError:
        return None


def latest_git_commit_date_local(repo_root: Path, rel_path: str):
    try:
        result = subprocess.run(
            ["git", "-C", str(repo_root), "log", "-1", "--format=%cI", "--", rel_path],
            check=False,
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            return None
        date_string = result.stdout.strip()
        if not date_string:
            return None
        dt = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
        return dt.astimezone(timezone.utc)
    except Exception:
        return None


def latest_mtime_date_local(repo_root: Path, rel_path: str):
    try:
        ts = (repo_root / rel_path).stat().st_mtime
        return datetime.fromtimestamp(ts, tz=timezone.utc)
    except Exception:
        return None


def main():
    local_base = Path(PAGES_LOCAL_PATH).expanduser().resolve() if PAGES_LOCAL_PATH else None
    use_local = bool(local_base and local_base.exists())

    def fetch_contents(path: str):
        url = f"{CODEBERG_API_BASE}/repos/{PAGES_OWNER}/{PAGES_REPO}/contents/{path}?ref={PAGES_BRANCH}"
        return fetch_json(url)

    projects_index = []
    logbook_index = []
    if use_local:
        try:
            projects_index = read_local_json(local_base, f"{PROJECTS_DIR}/index.json")
        except Exception:
            projects_index = []
        try:
            logbook_index = read_local_json(local_base, f"{LOGBOOK_DIR}/index.json")
        except Exception:
            logbook_index = []
    else:
        try:
            projects_index = fetch_json(f"{RAW_BASE}/{PROJECTS_DIR}/index.json")
        except Exception:
            projects_index = []
        try:
            logbook_index = fetch_json(f"{RAW_BASE}/{LOGBOOK_DIR}/index.json")
        except Exception:
            logbook_index = []

    if not isinstance(projects_index, list) or not projects_index:
        if use_local:
            try:
                contents = (local_base / PROJECTS_DIR).iterdir()
                projects_index = [p.name for p in contents if p.is_file()]
            except Exception:
                projects_index = projects_index if isinstance(projects_index, list) else []
        else:
            try:
                contents = fetch_contents(PROJECTS_DIR)
                projects_index = [item.get("name") for item in contents if item.get("type") == "file"]
            except Exception:
                projects_index = projects_index if isinstance(projects_index, list) else []

    if not isinstance(logbook_index, list) or not logbook_index:
        if use_local:
            try:
                contents = (local_base / LOGBOOK_DIR).iterdir()
                logbook_index = [p.name for p in contents if p.is_file()]
            except Exception:
                logbook_index = logbook_index if isinstance(logbook_index, list) else []
        else:
            try:
                contents = fetch_contents(LOGBOOK_DIR)
                logbook_index = [item.get("name") for item in contents if item.get("type") == "file"]
            except Exception:
                logbook_index = logbook_index if isinstance(logbook_index, list) else []

    logbook_files = [name for name in logbook_index if is_iso_date_md(name)] if isinstance(logbook_index, list) else []
    logbook_files.sort(reverse=True)

    project_files = (
        [name for name in projects_index if str(name).lower().endswith(".md") and not is_hidden_like(name)]
        if isinstance(projects_index, list)
        else []
    )

    projects_count = len(project_files)
    logbook_count = len(logbook_files)

    latest_logbook_file = logbook_files[0] if logbook_files else None

    project_candidates = []
    for name in project_files:
        repo_path = f"{PROJECTS_DIR}/{name}"
        date = None
        if use_local:
            date = latest_git_commit_date_local(local_base, repo_path) or latest_mtime_date_local(local_base, repo_path)
        else:
            try:
                date = latest_commit_date_for_path(repo_path)
            except Exception:
                date = None
        project_candidates.append({"name": name, "date": date})
    project_candidates.sort(
        key=lambda item: (
            -(item["date"].timestamp() if item["date"] else 0),
            str(item["name"]).lower(),
        )
    )

    latest_project_file = project_candidates[0]["name"] if project_candidates else None

    latest_project_title = "—"
    latest_project_desc = ""
    latest_project_link = ""
    if latest_project_file:
        rel = f"{PROJECTS_DIR}/{latest_project_file}"
        md = read_local_text(local_base, rel) if use_local else fetch_text(f"{RAW_BASE}/{rel}")
        latest_project_title = extract_h1(md) or latest_project_file.replace(".md", "")
        latest_project_desc = to_one_sentence(extract_first_paragraph(md))
        latest_project_link = link_to_pages_md(rel)

    latest_logbook_title = "—"
    latest_logbook_desc = ""
    latest_logbook_link = ""
    if latest_logbook_file:
        rel = f"{LOGBOOK_DIR}/{latest_logbook_file}"
        md = read_local_text(local_base, rel) if use_local else fetch_text(f"{RAW_BASE}/{rel}")
        date = latest_logbook_file.replace(".md", "")
        subtitle = extract_logbook_subtitle(md)
        latest_logbook_title = f"{date} — {subtitle}" if subtitle else date
        latest_logbook_desc = to_one_sentence(extract_logbook_body_excerpt(md))
        latest_logbook_link = link_to_pages_md(rel)

    clean_project_desc = html.escape(latest_project_desc).strip()
    clean_logbook_desc = html.escape(latest_logbook_desc).strip()

    new_block = f"""<!-- AUTO-GENERATED:START (do not edit by hand) -->
<div align="center">
<h2>Latest highlights</h2>

<table width="100%">
  <tr>
    <td width="50%" valign="top" align="center">
      <h3>Latest project writeup</h3>
      <p>{render_strong_link(latest_project_title, latest_project_link)}</p>
      <p>{clean_project_desc}</p>
    </td>
    <td width="50%" valign="top" align="center">
      <h3>Latest logbook entry</h3>
      <p>{render_strong_link(latest_logbook_title, latest_logbook_link)}</p>
      <p>{clean_logbook_desc}</p>
    </td>
  </tr>
</table>

<p><strong>Total project writeups</strong>: {projects_count} • <strong>Total logbook entries</strong>: {logbook_count}</p>

<p align="center"><small>Auto-updated from my <a href="https://codeberg.org/ankhseraph/pages">pages repo</a>.</small></p>
</div>
<!-- AUTO-GENERATED:END -->"""

    update_readme_generated_block(new_block)


if __name__ == "__main__":
    main()
